staging/src/k8s.io/apiserver/pkg/storage/cacher/store_btree.go

```diff
@@ -0,0 +1,404 @@
+	}
+
+	s.tree.AscendGreaterOrEqual(&storeElement{Key: continueKey}, func(i btree.Item) bool {
+		elementKey := i.(*storeElement).Key
+		if !strings.HasPrefix(elementKey, prefix) {
+			return false
+		}
+		// TODO: Might be worth to lookup one more item to provide more accurate HasMore.
+		if len(result) < limit {
+			result = append(result, i.(interface{}))
+		} else {
+			hasMore = true
+			return false
+		}
+		return true
+	})
+
+	return result, hasMore
+}
+
+func (s *btreeStore) Count(prefix, continueKey string) (count int) {
+	if continueKey == "" {
+		continueKey = prefix
+	}
+	s.tree.AscendGreaterOrEqual(&storeElement{Key: continueKey}, func(i btree.Item) bool {
+		elementKey := i.(*storeElement).Key
+		if !strings.HasPrefix(elementKey, prefix) {
+			return false
+		}
+		count++
+		return true
+	})
+
+	return count
+}
+
+func newIndexer(indexers cache.Indexers) indexer {
+	return indexer{
+		indices:  map[string]map[string]map[string]*storeElement{},
+		indexers: indexers,
+	}
+}
+
+type indexer struct {
+	indices  map[string]map[string]map[string]*storeElement
+	indexers cache.Indexers
+}
+
+func (i *indexer) ByIndex(indexName, indexValue string) ([]interface{}, error) {
+	indexFunc := i.indexers[indexName]
+	if indexFunc == nil {
+		return nil, fmt.Errorf("index with name %s does not exist", indexName)
+	}
+
+	index := i.indices[indexName]
+
+	set := index[indexValue]
+	list := make([]interface{}, 0, len(set))
+	for _, obj := range set {
+		list = append(list, obj)
+	}
+
+	return list, nil
+}
+
+func (i *indexer) Replace(objs []interface{}, resourceVersion string) error {
+	i.indices = map[string]map[string]map[string]*storeElement{}
+	for _, obj := range objs {
+		storeElem, ok := obj.(*storeElement)
+		if !ok {
+			return fmt.Errorf("obj not a storeElement: %#v", obj)
+		}
+		err := i.updateElem(storeElem.Key, nil, storeElem)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (i *indexer) updateElem(key string, oldObj, newObj *storeElement) (err error) {
+	var oldIndexValues, indexValues []string
+	for name, indexFunc := range i.indexers {
+		if oldObj != nil {
+			oldIndexValues, err = indexFunc(oldObj)
+		} else {
+			oldIndexValues = oldIndexValues[:0]
+		}
+		if err != nil {
+			return fmt.Errorf("unable to calculate an index entry for key %q on index %q: %w", key, name, err)
+		}
+
+		if newObj != nil {
+			indexValues, err = indexFunc(newObj)
+		} else {
+			indexValues = indexValues[:0]
+		}
+		if err != nil {
+			return fmt.Errorf("unable to calculate an index entry for key %q on index %q: %w", key, name, err)
+		}
+
+		index := i.indices[name]
+		if index == nil {
+			index = map[string]map[string]*storeElement{}
+			i.indices[name] = index
+		}
+
+		if len(indexValues) == 1 && len(oldIndexValues) == 1 && indexValues[0] == oldIndexValues[0] {
+			// We optimize for the most common case where indexFunc returns a single value which has not been changed
+			i.add(key, indexValues[0], newObj, index)
```

Comment:
This isn't needed - in this case the object changed by its index values didn't, so no need to update anything.

Reply:
We need to update the object.